src/libraries/System.Linq/src/System/Linq/Concat.cs

```diff
@@ -12,24 +12,52 @@ public static IEnumerable<TSource> Concat<TSource>(this IEnumerable<TSource> fir
         {
             if (first == null)
             {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.first);
+                ThrowHelper.ThrowArgumentNullException(nameof(first));
             }
 
             if (second == null)
             {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.second);
+                ThrowHelper.ThrowArgumentNullException(nameof(second));
             }
 
             return first is ConcatIterator<TSource> firstConcat
                 ? firstConcat.Concat(second)
-                : new Concat2Iterator<TSource>(first, second);
+                : new BaseConcatIterator<TSource>(first, second, rest: null);
         }
 
         /// <summary>
-        /// Represents the concatenation of two <see cref="IEnumerable{TSource}"/>.
+        /// Concatenates two or more sequences.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of the input sequences.</typeparam>
+        /// <param name="first">The first sequence to concatenate.</param>
+        /// <param name="second">The sequence to concatenate to the first sequence.</param>
+        /// <param name="rest">Specifies any subsequent sequences to concatenate.</param>
+        /// <returns>An <see cref="IEnumerable{T}"/> that contains the concatenated elements of the two input sequences.</returns>
+        public static IEnumerable<TSource> Concat<TSource>(this IEnumerable<TSource> first, IEnumerable<TSource> second, params IEnumerable<TSource>[] rest)
+        {
+            if (first == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(nameof(first));
+            }
+
+            if (second == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(nameof(second));
+            }
+
+            if (rest == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(nameof(rest));
+            }
+
+            return new BaseConcatIterator<TSource>(first, second, rest);
```

Comment:
Rather than all the other changes being made in the Concat implementation, why not just keep this simple and have it be:
```C#
IEnumerable<TSource> result = first.Concat(second);
foreach (IEnumerable<TSource> next in rest)
{
    result = result.Concat(next);
}
```
and have the rest of the concat implementation sort out the chaining?

Reply:
I was fairly split about what approach to follow here. Ultimately I went with modifying the implementation for a couple of reasons:

1. It doesn't fundamentally change the implementation characteristics of the existing Concat methods.
2. It avoids creating a linked list of nested concat iterators: this means reduced allocations plus it avoids _O(k^2)_ iteration of _k_ concatenated enumerables. In most applications that _k_ is constant though.